uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/PortletPrefsRESTController.java

```diff
@@ -0,0 +1,679 @@
+        } catch (IllegalArgumentException e) {
+            if (e.getMessage().contains("No portlet definition found for id '")) {
+                String error = "ERROR: User not authorized to access portlet '" + fname + "'.";
+                logger.info(error);
+                return new ResponseEntity<>(error, HttpStatus.FORBIDDEN);
+            }
+
+            logger.error(e.getMessage(), e);
+            return new ResponseEntity<>(e.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
+        } catch (Exception e) {
+
+            logger.error(e.getMessage(), e);
+            return new ResponseEntity<>(e.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    // get the entity preferences by comparing the entity with the defintion and returning the
+    // entity ones that are different
+    private Map<String, String[]> getEntityOnly(
+            Map<String, String[]> entity, Map<String, String[]> definition) {
+        Map<String, String[]> result = new HashMap<>();
+        for (Map.Entry<String, String[]> entry : entity.entrySet()) {
+            String key = entry.getKey();
+            String[] value = entry.getValue();
+            if (!definition.containsKey(key)) {
+                // the entity has something the definition does not
+                result.put(key, value);
+            } else {
+                // they both have it so compare the two
+                String[] defvalues = definition.get(key);
+                if (value == null) {
+                    if (defvalues != null) {
+                        result.put(key, value);
+                    } // otherwise they're both null so don't add
+                } else {
+                    if (!Arrays.equals(value, defvalues)) {
+                        result.put(key, value);
+                    } // otherwise they're the same so don't add
+                }
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Get uPortal definition preferecnces for portlet/soffit/webcomponent. This method provides a
+     * REST interface for uPortal for obtaining the definition preferences for the
+     * portlet/soffit/webcomponent
+     *
+     * <p>The path for this method is /prefs/getdefinitionprefs/fname. The fname is a string
+     * representing the fname of the portlet/soffit/webcomponent to return the composite definition
+     * preferences. These are the default preferences for the given portlet/soffit/webcomponent
+     * including any non-overwritten descriptor preferences for the given
+     * portlet/soffit/webcomponent. Will include extra preferences for soffits including the url, so
+     * please set any of these to read-only when you define them. You must be logged in as a portal
+     * administrator or have configuration rights for the given portlet/soffit/webcomponent in
+     * uPortal to access these. Will return a 403 or 404 otherwise
+     */
+    // must be authenticated and a portal administrator
```

Comment:
This comment should not be here or like that.

Reply:
I missed that one. I'll remove it